drivers: w1: Make 1-Wire Skip ROM cmd. optional

hubpav opened the pull request:
For systems with a true dynamic 1-Wire nature, the 1-Wire devices cannot be defined in the DeviceTree (the number of slaves in the system is not known during the build time).

In other words, there are no pre-defined 1-Wire devices on the bus in DeviceTree, and thus the slave_count variable is always zero.

That means the skip_rom functionality will always be called even if there are multiple devices connected to the bus.

This commit allows the user to override this original behavior and avoid the skip_rom call.

Also, this work preserves full backward compatibility.

Signed-off-by: Pavel Hübner <pavel.hubner@hardwario.com>

dp7hgh7 commented:
In order make mare clear what this functionality really does, I suggest to invert it and call the config option something like `W1_IGNORE_SLAVE COUNT`.

amirfuhrmann commented:
> In order make mare clear what this functionality really does, I suggest to invert it and call the config option something like `W1_IGNORE_SLAVE COUNT`.

Much better. Thanks

carlescufi commented:
@str4t0m can you take a look?

carlescufi commented on drivers/w1/w1_net.c:
```diff
@@ -346,11 +346,15 @@ int w1_skip_rom(const struct device *dev, const struct w1_slave_config *config)
+#if W1_IGNORE_SLAVE_COUNT
+	return match_rom(dev, config);
+#else
 	if (w1_get_slave_count(dev) > 1) {
```
First, you need to prefix with `CONFIG_`.
Also, you could do instead:

```
if (!IS_ENABLED(CONFIG_W1_IGNORE_SLAVE_COUNT) && w1_get_slave_count(dev) > 1) {
```

hubpav replied:
True, done.

str4t0m reviewed (changes requested):
Hi, thanks for this proposal.
I am not yet sure if this is the best way to solve this issue.
I need to consider a few other use cases, an will comment in the next days.

str4t0m commented on drivers/w1/Kconfig:
```diff
@@ -40,6 +40,15 @@ config W1_SHELL_BUFFER_SIZE
 	  To be able to write more than 8 bytes make
 	  sure to also increase SHELL_ARGC_MAX accordingly.
 
+config W1_IGNORE_SLAVE_COUNT
```
Ignore_slave_count is not really descriptive what it is really doing, and could be understood to have exactly the opposite effect.
How about calling it something like `W1_NET_FORCE_MULTIDROP_ADDRESSING`, highlighting in the description that usually this does not need to be enabled.

hubpav replied:
Good point - I have changed the symbol name and updated the help text accordingly.

str4t0m reviewed (changes requested):
Please use a more descriptive name then we can add this option.

CodingGhost commented:
I just stumbled across the same issue. This would really help.

hubpav committed:
drivers: w1: Make 1-Wire Skip ROM cmd. optional

For systems with a true dynamic 1-Wire nature, the 1-Wire devices
cannot be defined in the DeviceTree (the number of slaves in
the system is not known during the build time).

In other words, there are no pre-defined 1-Wire devices on the bus in
DeviceTree, and thus the slave_count variable is always zero.

That means the skip_rom functionality will always be called even if
there are multiple devices connected to the bus.

This commit allows the user to override this original behavior and avoid
the skip_rom call by ignoring the slave count variable.

Also, this work preserves full backward compatibility.

Signed-off-by: Pavel Hübner <pavel.hubner@hardwario.com>

hubpav commented on drivers/w1/Kconfig:
```diff
@@ -40,6 +40,15 @@ config W1_SHELL_BUFFER_SIZE
 	  sure to also increase SHELL_ARGC_MAX accordingly.
 
+config W1_IGNORE_SLAVE_COUNT
+	bool "Ignore 1-Wire slave count"
```
Done

hubpav commented:
@carlescufi Hi. The requested changes have been implemented. Would you please review and approve the changes? Thx a lot!

carlescufi commented:
> @carlescufi Hi. The requested changes have been implemented. Would you please review and approve the changes? Thx a lot!

Promin za zpozdeni!

hubpav commented:
@carlescufi No worries. We are happy that this got merged!
